### ISSUE TITLE
Scoping issue with switch and let statements

### ISSUE THREAD

**Issue by izderadicka** [babel/babel#5378]
Hi I have found this problem:

### Input Code

```js
let myVar = {val:1};

switch (myVar.val) {
    case 0:
        alert('MyVar.val is 0');
        break;
    case 1:
        let myVar = 'dummy';
        alert('Myvar.val is 1');
}
```

### Babel Configuration (.babelrc, package.json, cli command)
Using env preset - everything pretty much as in CLI tutorial

.babelrc:
```js
{
  "presets": ["env"]
}
```

### Expected Behavior
Above code works fine in recent firefox and chromium

### Current Behavior

babel transpilation is
```js
'use strict';

var myVar = { val: 1 };

switch (_myVar.val) {
    case 0:
        alert('MyVar.val is 0');
        break;
    case 1:
        var _myVar = 'dummy';
        alert('Myvar.val is 1');
}
```

This causes undefined variable error.

Clearly there is some misunderstanding about scope of variable declared in switch statement scope.
While browsers use in switch(expression)   variable from outer scope,  babel is here using variable from switch block scope.

### Possible Solution
Modify behavior of babel to be consistent with current browser -  e.g. switch(expression)  variable scope is the outer scope.

### Your Environment

| software         | version
| ---------------- | -------
| Babel            |  6.23.0 (babel-core 6.23.1)
| node             | v6.9.5
| npm              | 3.10.10
| Operating System |  Linux Ubuntu 16.04 64 bits

**Comment by xtuc** [comment]
The `case` is just a list of statements and isn't providing a real scope.

The following code works:

```js
switch (myVar.val) {
    case 0:
        alert('MyVar.val is 0');
        break;
    case 1: {
        let myVar = 'dummy';
        alert('Myvar.val is 1');
    }
}
```

[Try in REPL](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=true&presets=es2015&targets=Node-4&browsers=&builtIns=false&code=let%20myVar%20%3D%20%7Bval%3A1%7D%3B%0A%0Aswitch%20(myVar.val)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20%20%20alert('MyVar.val%20is%200')%3B%0A%20%20%20%20%20%20%20%20break%3B%0A%20%20%20%20case%201%3A%20%7B%0A%20%20%20%20%20%20%20%20let%20myVar%20%3D%20'dummy'%3B%0A%20%20%20%20%20%20%20%20alert('Myvar.val%20is%201')%3B%0A%20%20%20%20%7D%0A%7D)

**Comment by aaronang** [comment]
I found the following. I think the code is compiled correctly in:
* ES2016, [try in REPL](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2016&targets=&browsers=&builtIns=false&code=let%20v%20%3D%200%3B%0A%0Aswitch%20(v)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20let%20v%20%3D%201%3B%0A%7D).
* ES2017, [try in REPL](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2017&targets=&browsers=&builtIns=false&code=let%20v%20%3D%200%3B%0A%0Aswitch%20(v)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20let%20v%20%3D%201%3B%0A%7D).

But not in ES2015, [try in REPL](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2015&targets=&browsers=&builtIns=false&code=let%20v%20%3D%200%3B%0A%0Aswitch%20(v)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20let%20v%20%3D%201%3B%0A%7D).

**Comment by xtuc** [comment]
@aaronang With ES2015 we need to transpile `let` to `var` and emulate a scoping in some cases. This issue occurs during the transformation.

**Comment by aaronang** [comment]
@xtuc But if you look at the [REPL](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2015&targets=&browsers=&builtIns=false&code=let%20v%20%3D%200%3B%0A%0Aswitch%20(v)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20let%20v%20%3D%201%3B%0A%7D) of `babel-preset-es2015`. The compiled code is invalid too.

**Comment by yavorsky** [comment]
BTW, Wrapping `case` into the block solves the error:

```js
let v = 0;

switch (v) {
    case 0: {
      let v = 1;
    }
}
```

**Comment by izderadicka** [comment]
[Adding block in case](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=env&targets=&browsers=&builtIns=false&code=let%20v%20%3D%200%3B%0A%0Aswitch%20(v)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20let%20v%20%3D%201%3B%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%0A%7D) indeed helps.   But as I understood all case statements are by default in one  block - thanks to optional break statement (as discussed [here](https://esdiscuss.org/topic/let-const-in-switch-cases) or shown on MDN [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#Temporal_dead_zone_and_errors_with_let)).  This is correctly recognized as shown [here](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=env&targets=&browsers=&builtIns=false&code=let%20v%20%3D%200%3B%0A%0Aswitch%20(v)%20%7B%0A%20%20%20%20case%200%3A%0A%20%20%20%20%20%20let%20v%20%3D%201%3B%0A%20%20%20%20%20%20%0A%20%20%20%20case%201%3A%0A%20%20%20%20%20%20let%20v%20%3D%202%3B%0A%7D).   

So the problem is about scope of expression in switch() statement in in transpling to 2015.

**Comment by loganfsmyth** [comment]
I agree that this appears to be a bug. The curlies in `switch(x) { ... }` introduce their own block scope, so the argument to `switch` should not be transformed.

**Comment by xtuc** [comment]
The issue seems to come from `scope.rename` ([Renamer](https://github.com/babel/babel/blob/7.0/packages/babel-traverse/src/scope/lib/renamer.js)). Which in the following example also rename the switch discriminant:

```js
let v = 0;

switch (v) {
    case 0:
      let v = 1; // <- rename v in the scope from there
}
```

The `rename` method isn't well tested, in did add some tests there: https://github.com/xtuc/babel/blob/45c6808415e5454162b3771b2c9e65d199ab5f2f/packages/babel-traverse/test/scope.js#L32-L58. Currently failing because it reproduce this issue.

I wasn't able to resolve this issue.

**Comment by liuxingbaoyu** [comment]
This appears to have been fixed.
